File: python/examples/payment_link_server.py
"""Minimal payment link server for interop testing.

Listens on port 3004. Requires Surfpool on localhost:8899.
"""
from __future__ import annotations

import asyncio
import json
import os
import random
from http.server import BaseHTTPRequestHandler, HTTPServer

from solana.rpc.async_api import AsyncClient

from solana_mpp._headers import format_www_authenticate, parse_authorization
from solana_mpp.server.mpp import ChargeOptions, Config, Mpp
from solana_mpp.server.payment_page import (
    accepts_html,
    challenge_to_html,
    is_service_worker_request,
    service_worker_js,
)
from solana_mpp.store import MemoryStore

RECIPIENT = "CXhrFZJLKqjzmP3sjYLcF4dTeXWKCy9e2SXXZ2Yo6MPY"
USDC_MINT = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"
SECRET = "interop-test-secret-key-long-enough-for-hmac-sha256-operations"
RPC_URL = os.environ.get("RPC_URL", "http://localhost:8899")

FORTUNES = [
    "A smooth long journey!",
    "Good news will come to you by mail.",
    "Curiosity kills boredom.",
]

mpp = Mpp(Config(
    recipient=RECIPIENT,
    secret_key=SECRET,
    currency=USDC_MINT,
    decimals=6,
    network="localnet",
    rpc_url=RPC_URL,
    html=True,
    store=MemoryStore(),
    rpc=AsyncClient(RPC_URL),
))

# Fund recipient at startup
try:
    import httpx

    httpx.post(RPC_URL, json={
        "jsonrpc": "2.0", "id": 1, "method": "surfnet_setAccount",
        "params": [RECIPIENT, {"lamports": 1_000_000_000, "data": "", "executable": False, "owner": "11111111111111111111111111111111", "rentEpoch": 0}],
    }, timeout=5)
    httpx.post(RPC_URL, json={
        "jsonrpc": "2.0", "id": 1, "method": "surfnet_setTokenAccount",
        "params": [RECIPIENT, USDC_MINT, {"amount": 0, "state": "initialized"}, "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA"],
    }, timeout=5)
except Exception:
    pass


class Handler(BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass  # Suppress request logging

    def do_GET(self):
        if self.path == "/health":
            self._json(200, {"ok": True})
            return

        if not self.path.startswith("/fortune"):
            self._json(404, {"error": "not found"})
            return

        # Service worker
        if is_service_worker_request(self.path):
            self.send_response(200)
            self.send_header("Content-Type", "application/javascript")
            self.send_header("Service-Worker-Allowed", "/")
            self.end_headers()
            self.wfile.write(service_worker_js().encode())
            return

        # Check for credential
        auth = self.headers.get("Authorization", "")
        if auth.startswith("Payment "):
            try:
                credential = parse_authorization(auth)
                receipt = asyncio.run(mpp.verify_credential(credential))
                fortune = random.choice(FORTUNES)
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.send_header("Payment-Receipt", receipt.reference)
                self.end_headers()
                self.wfile.write(json.dumps({"fortune": fortune}).encode())
                return
            except Exception as e:
                pass  # Fall through to challenge

        # Issue challenge
        challenge = mpp.charge_with_options("0.01", ChargeOptions(description="Open a fortune cookie"))
        www_auth = format_www_authenticate(challenge)

        if accepts_html(self.headers.get("Accept")):
            html = challenge_to_html(challenge, RPC_URL, "localnet")
            self.send_response(402)
            self.send_header("Content-Type", "text/html; charset=utf-8")
            self.send_header("WWW-Authenticate", www_auth)
            self.send_header("Cache-Control", "no-store")
            self.end_headers()
            self.wfile.write(html.encode())
        else:
            body = json.dumps({
                "type": "https://paymentauth.org/problems/payment-required",
                "title": "Payment Required",
                "status": 402,
            })
            self.send_response(402)
            self.send_header("Content-Type", "application/json")
            self.send_header("WWW-Authenticate", www_auth)
            self.send_header("Cache-Control", "no-store")
            self.end_headers()
            self.wfile.write(body.encode())

    def _json(self, status: int, data: dict):
        body = json.dumps(data).encode()
        self.send_response(status)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(body)


if __name__ == "__main__":
    port = 3004
    server = HTTPServer(("0.0.0.0", port), Handler)
    print(f"Python payment-link-server listening on http://localhost:{port}")
    server.serve_forever()

File: rust/examples/payment_link_server.rs
//! Minimal test server for Playwright E2E tests of the HTML payment link flow.
//!
//! Run with: `cargo run --example payment_link_server`
//! Listens on http://localhost:3001
//! Requires Surfpool on localhost:8899

use axum::{
    extract::Query,
    http::{header, HeaderMap, StatusCode},
    response::{IntoResponse, Response},
    routing::get,
};
use solana_mpp::protocol::core::headers::parse_authorization;
use solana_mpp::server::{html, Config, Mpp};
use std::collections::HashMap;
use std::sync::Arc;

const CSP: &str = "default-src 'self'; script-src 'unsafe-inline'; style-src 'unsafe-inline'; connect-src *; worker-src 'self'";

async fn fortune(
    headers: HeaderMap,
    Query(params): Query<HashMap<String, String>>,
    mpp: axum::extract::State<Arc<Mpp>>,
) -> Response {
    // Authenticated request — verify the credential and return fortune.
    if let Some(auth) = headers.get(header::AUTHORIZATION) {
        let auth_str = auth.to_str().unwrap_or("");
        if auth_str.starts_with("Payment ") {
            match parse_authorization(auth_str) {
                Ok(credential) => {
                    match mpp.verify_credential(&credential).await {
                        Ok(receipt) => {
                            let ref_id = receipt.reference;
                            return (
                                StatusCode::OK,
                                [
                                    ("content-type", "application/json"),
                                    ("payment-receipt", &ref_id),
                                ],
                                r#"{"fortune":"A smooth sea never made a skilled sailor."}"#,
                            )
                                .into_response();
                        }
                        Err(e) => {
                            eprintln!("verify_credential failed: {e}");
                            // Fall through to re-issue challenge
                        }
                    }
                }
                Err(e) => {
                    eprintln!("parse_authorization failed: {e}");
                }
            }
        }
    }

    // Service worker JS.
    if params.contains_key(html::SERVICE_WORKER_PARAM) {
        return (
            StatusCode::OK,
            [
                ("content-type", "application/javascript"),
                ("service-worker-allowed", "/"),
            ],
            html::service_worker_js(),
        )
            .into_response();
    }

    // Generate challenge.
    let challenge = mpp
        .charge_with_options(
            "0.01",
            solana_mpp::server::ChargeOptions {
                description: Some("Open a fortune cookie"),
                ..Default::default()
            },
        )
        .expect("charge");
    let www_auth = solana_mpp::format_www_authenticate(&challenge).unwrap_or_default();

    // Browser — HTML payment page.
    let accept = headers
        .get(header::ACCEPT)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("");
    if html::accepts_html(accept) {
        let page = html::challenge_to_html(&challenge, mpp.rpc_url(), mpp.network());
        return (
            StatusCode::PAYMENT_REQUIRED,
            [
                ("content-type", "text/html; charset=utf-8"),
                ("content-security-policy", CSP),
                ("www-authenticate", &www_auth),
            ],
            page,
        )
            .into_response();
    }

    // API client — JSON 402.
    (
        StatusCode::PAYMENT_REQUIRED,
        [
            ("content-type", "application/json"),
            ("www-authenticate", &www_auth),
        ],
        serde_json::to_string(&challenge).unwrap(),
    )
        .into_response()
}

async fn health() -> impl IntoResponse {
    (
        StatusCode::OK,
        [("content-type", "application/json")],
        r#"{"ok":true}"#,
    )
}

#[tokio::main]
async fn main() {
    let rpc_url = std::env::var("RPC_URL").unwrap_or_else(|_| "http://localhost:8899".to_string());

    // No fee payer — test mode client pays its own fees.
    let mpp = Arc::new(
        Mpp::new(Config {
            recipient: "CXhrFZJLKqjzmP3sjYLcF4dTeXWKCy9e2SXXZ2Yo6MPY".to_string(),
            secret_key: Some(
                "e2e-test-secret-key-long-enough-for-hmac-operations-1234567890".into(),
            ),
            network: "localnet".to_string(),
            rpc_url: Some(rpc_url),
            currency: "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v".to_string(),
            decimals: 6,
            html: true,
            ..Default::default()
        })
        .expect("valid config"),
    );

    // Fund the recipient so their token account exists (surfpool cheatcode).
    let rpc_url = mpp.rpc_url().to_string();
    let recipient = "CXhrFZJLKqjzmP3sjYLcF4dTeXWKCy9e2SXXZ2Yo6MPY";
    let mint = "EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v";
    let token_program = "TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA";

    // Set recipient SOL balance
    let _ = reqwest::Client::new()
        .post(&rpc_url)
        .json(&serde_json::json!({
            "jsonrpc": "2.0", "id": 1,
            "method": "surfnet_setAccount",
            "params": [recipient, {"lamports": 1_000_000_000_u64, "data": "", "executable": false, "owner": "11111111111111111111111111111111", "rentEpoch": 0}]
        }))
        .send().await;

    // Set recipient USDC token account
    let _ = reqwest::Client::new()
        .post(&rpc_url)
        .json(&serde_json::json!({
            "jsonrpc": "2.0", "id": 1,
            "method": "surfnet_setTokenAccount",
            "params": [recipient, mint, {"amount": 0, "state": "initialized"}, token_program]
        }))
        .send()
        .await;

    let app = axum::Router::new()
        .route("/fortune", get(fortune))
        .route("/health", get(health))
        .with_state(mpp);

    let listener = tokio::net::TcpListener::bind("0.0.0.0:3001").await.unwrap();
    println!("payment_link_server listening on http://localhost:3001");
    axum::serve(listener, app).await.unwrap();
}
